src/ok.rs

```diff
@@ -605,3 +697,235 @@ fn memlock_rlimit() -> TestResult {
         mesg: Some(mesg),
     }
 }
+
+fn msr_bit_read(reg: u32, cpu: u16) -> Result<u64, anyhow::Error> {
+    let mut msr = Msr::new(reg, cpu).context("Error Reading MSR")?;
+    let raw_value = msr.read()?;
+    Ok(raw_value)
+}
+
+fn sme_test() -> TestResult {
+    let raw_value = match msr_bit_read(0xC0010010, 0) {
+        Ok(raw) => raw,
+        Err(e) => {
+            return TestResult {
+                name: "SME".to_string(),
+                stat: TestState::Fail,
+                mesg: format!("MSR read failed: {}", e).into(),
+            }
+        }
+    };
+    let (testres, mesg) = match raw_value >> 23 & 1 {
+        1 => (TestState::Pass, "Enabled"),
+        0 => (TestState::Fail, "Disabled"),
+        _ => unreachable!(),
+    };
+    TestResult {
+        name: "SME".to_string(),
+        stat: testres,
+        mesg: Some(mesg.to_string()),
+    }
+}
+
+fn snp_test() -> TestResult {
+    let raw_value = match msr_bit_read(0xC0010010, 0) {
+        Ok(raw) => raw,
+        Err(e) => {
+            return TestResult {
+                name: "SNP".to_string(),
+                stat: TestState::Fail,
+                mesg: format!("MSR read failed: {}", e).into(),
+            }
+        }
+    };
+    let (testres, mesg) = match raw_value >> 24 & 1 {
+        1 => (TestState::Pass, "Enabled"),
+        0 => (TestState::Fail, "Disabled"),
+        _ => unreachable!(),
+    };
+    TestResult {
+        name: "SNP".to_string(),
+        stat: testres,
+        mesg: Some(mesg.to_string()),
+    }
+}
+
+fn get_rmp_address() -> TestResult {
+    let rmp_base = match msr_bit_read(0xC0010132, 0) {
+        Ok(raw) => raw,
+        Err(e) => {
+            return TestResult {
+                name: "Reading RMP table".to_string(),
+                stat: TestState::Fail,
+                mesg: format!("Failed to read the desired MSR: {}", e).into(),
+            }
+        }
+    };
+    let rmp_end = match msr_bit_read(0xC0010133, 0) {
+        Ok(raw) => raw,
+        Err(e) => {
+            return TestResult {
+                name: "Reading RMP table".to_string(),
+                stat: TestState::Fail,
+                mesg: format!("Failed to read the desired MSR: {}", e).into(),
+            }
+        }
+    };
+    if rmp_base == 0 || rmp_end == 0 {
+        TestResult {
+            name: "Reading RMP table".to_string(),
+            stat: TestState::Fail,
+            mesg: format!("RMP table was not read successfully").into(),
+        }
+    } else {
+        TestResult {
+            name: "RMP table addresses".to_string(),
+            stat: TestState::Pass,
+            mesg: format!("Addresses: {} - {}", rmp_base, rmp_end).into(),
+        }
+    }
+}
+
+fn snp_ioctl(test: SnpIoctl) -> TestResult {
+    let status = match snp_platform_status() {
+        Ok(stat) => stat,
+        Err(e) => {
+            return TestResult {
+                name: "Looking at SNP platform status".to_string(),
```

Comment:
You can use your enum display implementation to show what is the test that is failing instead of this generic message.